go/pkg/cas/upload.go

```diff
@@ -589,6 +601,143 @@ func (u *uploader) uploadBatch(ctx context.Context, reqs []*repb.BatchUpdateBlob
 	})
 }
 
+// stream uploads the item using ByteStream service.
+//
+// If the blob is already uploaded, then the function returns quickly and
+// without an error.
+func (u *uploader) stream(ctx context.Context, item *uploadItem, updateCacheStats bool) error {
+	if err := u.semByteStreamWrite.Acquire(ctx, 1); err != nil {
+		return err
+	}
+	defer u.semByteStreamWrite.Release(1)
+
+	// Allocate resources.
+	enc := zstdEncoders.Get().(*zstd.Encoder)
+	fileBuf := u.fileIOBufs.Get().([]byte)
+	streamBuf := u.streamBufs.Get().([]byte)
+	defer func() {
+		zstdEncoders.Put(enc)
+		u.fileIOBufs.Put(fileBuf)
+		u.streamBufs.Put(streamBuf)
+	}()
+
+	// Open the item.
+	r, err := item.Open()
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+
+	// TODO(nodir): implement per-RPC timeouts. No nice way to do it.
+	rewind := false
+	return u.withRetries(ctx, func(ctx context.Context) error {
+		// Do not rewind if this is the first attempt.
+		if rewind {
+			if _, err := r.Seek(0, io.SeekStart); err != nil {
+				return err
+			}
+		}
+		rewind = true
+
+		// Start the byte stream.
+		stream, err := u.byteStream.Write(ctx)
+		if err != nil {
+			return err
+		}
+
+		// Compress using an in-memory pipe. This is mostly to accomodate the fact
+		// that zstd package expects a writer.
+		// Note that using io.Pipe() means we buffer only bytes that were not uploaded yet.
+		pr, pw := io.Pipe()
+		enc.Reset(pw)
+
+		// Read from disk and make RPCs concurrently.
+		eg, ctx := errgroup.WithContext(ctx)
+		eg.Go(func() error {
+			switch _, err := io.CopyBuffer(enc, r, fileBuf); {
+			case err == io.ErrClosedPipe:
+				// The other goroutine exited before we finished encoding.
+				// Might be a cache hit or context cancelation.
+				// In any case, the other goroutine has the actual error, so return nil
+				// here.
+				return nil
+			case err != nil:
+				return errors.Wrapf(err, "failed to read the file/blob")
+			}
```

Comment:
It looks like `enc.Close()` will not be called if we return in either of these two cases.

Reply:
done